### PR TITLE
Configure persistent datastore index settings

### PR DESCRIPTION
**Pull request by b-deam**
With this commit we add the ability for users to specify the number of
primary and replica shard counts for `rally-*` indices when using a 
persistent metrics store through the configuration of `reporting` settings 
in `rally.ini`.

Closes https://github.com/elastic/rally/issues/1309

### CODE REVIEW

File: docs/configuration.rst [hunk 1]
```diff
@@ -69,6 +69,8 @@ The following settings are applicable only if ``datastore.type`` is set to "elas
 * ``datastore.user``: Sets the name of the Elasticsearch user for the metrics store.
 * ``datastore.password``: Sets the password of the Elasticsearch user for the metrics store.
 * ``datastore.probe.cluster_version`` (default: true): Enables automatic detection of the metric store's version.
+* ``datastore.number_of_shards`` (default: 1): The number of primary shards that the ``rally-*`` indices should have. Any updates to this setting after initial index creation will only be applied to new ``rally-*`` indices.
+* ``datastore.number_of_replicas`` (default: 0): The number of replicas each primary shard has. Defaults to 0. Any updates to this setting after initial index creation will only be applied to new ``rally-*`` indices.
```
Comment (dliappis) [inline docs/configuration.rst:73]: We should also mention in `migrate.rst` that starting from version ... it's possible to configure the number of shards and replicas of Rally metrics indices, when an Elasticsearch store is in use.

File: esrally/metrics.py [hunk 2]
```diff
@@ -250,7 +251,22 @@ def results_template(self):
...
+            # if we're using a persistent metrics store, then we should check and update the index templates settings
+            if self.cfg.opts("reporting", "datastore.type") == "elasticsearch":
+                return self._update_template(f)
+
+            else:
+                return f.read()
+
+    def _update_template(self, template_name):
+        template = json.load(template_name)
+        template["settings"]["index"]["number_of_shards"] = self.cfg.opts(
```
Comment (dliappis) [inline esrally/metrics.py:263]: somehow, referencing config parameters directly here feels (to me) slightly out of place.
I'd have defined them as instance variables instead e.g. `self._number_of_shards = self.cfg.opts("reporting", "datastore.number_of_shards", default_value=1, mandatory=False)` and `self._number_of_replicas = ...` for better readability and discoverability but obviously what you have here works fine too.

File: esrally/metrics.py [hunk 3]
```diff
@@ -250,7 +251,22 @@ def results_template(self):
 
     def _read(self, template_name):
         with open("%s/resources/%s.json" % (self.script_dir, template_name), encoding="utf-8") as f:
-            return f.read()
+            # if we're using a persistent metrics store, then we should check and update the index templates settings
+            if self.cfg.opts("reporting", "datastore.type") == "elasticsearch":
```
Comment (dliappis) [inline esrally/metrics.py:255]: This class doesn't get called when there is an in-memory metrics store; the references are always from EsXYZStore i.e.  https://github.com/elastic/rally/blob/ede9670d50f1d0330baf09738a01e6535b17a49f/esrally/metrics.py#L862 / https://github.com/elastic/rally/blob/ede9670d50f1d0330baf09738a01e6535b17a49f/esrally/metrics.py#L1607 / https://github.com/elastic/rally/blob/ede9670d50f1d0330baf09738a01e6535b17a49f/esrally/metrics.py#L1700

So we don't need this conditional. Therefore, we can also move the code inside `_update_template()` here and reference the shard settings from instance variables, see the comment below.

**Review by dliappis — COMMENTED** [review]
This looks good and works fine in tests. Left a few comments.

File: esrally/config.py [hunk 4]
```diff
@@ -210,7 +210,7 @@ def exists(self, section, key):
 
     def config_present(self):
         """
-        :return: true iff a config file already exists.
+        :return: true if a config file already exists.
```
[the remark this answers is not in the document]
Comment (b-deam) [reply esrally/config.py:213]: Wow, TIL.

File: esrally/metrics.py [hunk 5]
```diff
@@ -238,6 +238,7 @@ class IndexTemplateProvider:
 
     def __init__(self, cfg):
         self.script_dir = cfg.opts("node", "rally.root")
+        self.cfg = cfg
```
[the remark this answers is not in the document]
Comment (b-deam) [reply esrally/metrics.py:241]: I agree, `self._config` seems better fitting given the config is really only ever intended to be used privately.

File: esrally/metrics.py (hunk 3 above)
Comment (b-deam) [reply esrally/metrics.py:255]: Good pick up!

**Comment by b-deam** [comment]
I made the changes - but given that it _shouldn't_ be possible for an instance of `InMemoryMetricsStore` to have an `index_template_provider_class` attribute I wasn't sure on how to test this.

**Comment by dliappis** [comment]
> I made the changes - but given that it _shouldn't_ be possible for an instance of `InMemoryMetricsStore` to have an `index_template_provider_class` attribute I wasn't sure on how to test this.

Additional settings like `datastore.host` simply get ignored when `datastore.type != elasticsearch` so I am not sure what you are aiming to test?

**Comment by b-deam** [comment]
> > I made the changes - but given that it _shouldn't_ be possible for an instance of `InMemoryMetricsStore` to have an `index_template_provider_class` attribute I wasn't sure on how to test this.
> 
> Additional settings like `datastore.host` simply get ignored when `datastore.type != elasticsearch` so I am not sure what you are aiming to test?

Right - I guess I want to ensure that if any changes are made to the `InMemoryMetricsStore` in so that if it ever implements/utilises a `index_template_provider_class`, then I guess we should make sure that we catch that in unit tests because all of the logic around  is currently based on the assumption that it will only ever be executed in the context of a `EsMetricsStore`.

**Comment by dliappis** [comment]
>  if any changes are made to the `InMemoryMetricsStore` in so that if it ever implements/utilises a `index_template_provider_class`

I don't think this will ever happen. @danielmitterdorfer do you feel we need a unit test for this unlikely scenario?

**Comment by danielmitterdorfer** [comment]
> I don't think this will ever happen. @danielmitterdorfer do you feel we need a unit test for this unlikely scenario?

No, that will not happen. The index template provider is designed to be used only with an `EsMetricsStore`.

**Comment by b-deam** [comment]
Thanks @danielmitterdorfer, @dliappis. 

I had the misconception that 'In-memory' == local metrics store - so it's obvious now that we'd never implement a template provider 🤦 .

I removed the unit test for the `InMemoryMetricsStore`. Will merge after CI passes.

**Review by dliappis — APPROVED** [review]
LGTM

**Pull request merged**